workflows/genomes_annotation.nf

```diff
@@ -134,17 +138,100 @@ workflow GAP {
     MASH_TO_NWK(
         dereplicated_genomes.out.mash_splits | flatten
     )
+    
+    GTDBTK_QC(
+        dereplicated_genomes.out.single_genomes_fna_tuples.map({ it[1] }) \
+        .mix(dereplicated_genomes.out.many_genomes_fna_tuples.filter {
+            it[1].name.contains(it[0])
+        }.map({ it[1] })) \
+        .collect(),
+        channel.value("fa"), // genome file extension
+        ch_gtdb_db
+    )
 
+    gtdbtk_tables_qc_ch = channel.empty() \
+        .mix(GTDBTK_QC.out.gtdbtk_summary_bac120, GTDBTK_QC.out.gtdbtk_summary_arc53) \
+        .collectFile(name: 'gtdbtk.summary.tsv')
+        
+    IDENTIFY_DOMAIN(
+        gtdbtk_tables_qc_ch
+    )
+    
+    accessions_with_domains_ch = IDENTIFY_DOMAIN.out.detected_domains.flatMap { file ->
+        file.readLines().collect { line ->
+            def (genomeName, domain) = line.split(',')
+            [genomeName, domain]
+        }
+    }
+            
     PROCESS_MANY_GENOMES(
-        dereplicated_genomes.out.many_genomes_fna_tuples
+        dereplicated_genomes.out.many_genomes_fna_tuples,
+        accessions_with_domains_ch
     )
 
     PROCESS_SINGLETON_GENOMES(
         dereplicated_genomes.out.single_genomes_fna_tuples,
         PREPARE_DATA.out.genomes_checkm.first(),
+        accessions_with_domains_ch,
         ch_gunc_db
     )
+    
+    GENERATE_COMBINED_QC_REPORT(
+        PREPARE_DATA.out.qs50_failed,
+        PROCESS_SINGLETON_GENOMES.out.gunc_failed_txt,
+        IDENTIFY_DOMAIN.out.detected_domains
+    )
 
+    // Separate accessions into those we don't know domain for (Undefined) and those that we do
+    accessions_with_domains_ch
+    .branch {
+        defined: it[1] != "Undefined"
+        return it[0]
+        undefined: it[1] == "Undefined"
+        return it[0]
+    }
+    .set { domain_splits }
+    
+    // Add "to_remove" to accessions that have an undefined domain
+    undefined_genomes = domain_splits.undefined.map(it -> [it, "to_remove"])
+    
+    // Do the same with genomes that were removed by GUNC (these are loaded from file first)
+    gunc_removed_genomes = PROCESS_SINGLETON_GENOMES.out.gunc_failed_txt.map { file_path ->
+        def contents = file(file_path).splitText().collect { line ->
+        line.trim()
+        }
+        return contents
+    }.flatten().map(it -> [it, "to_remove"])
+    
+    // Make a single set of tuples to remove, then remove these from fna tuples for singleton genomes
+    combined_list_to_remove = gunc_removed_genomes.concat(undefined_genomes)
+    filtered_single_genome_fna_tuples = dereplicated_genomes.out.single_genomes_fna_tuples \
+        .join(combined_list_to_remove, remainder: true) \
+        .filter { it -> it[2] == null} \
+        .map { it -> [it[0], it[1]] }
```

Comment:
Can you please use names instead of it[idx]? It helps making the code much more easy to read.

Reply:
Done in https://github.com/EBI-Metagenomics/genomes-catalogue-pipeline/commit/b297cfc3000403e8871ad7edb4a194d1a3739dcd